tls/s2n_client_hello.c

```diff
@@ -971,3 +972,30 @@ int s2n_client_hello_has_extension(struct s2n_client_hello *ch, uint16_t extensi
     }
     return S2N_SUCCESS;
 }
+
+int s2n_client_hello_get_supported_groups(struct s2n_client_hello *ch, uint16_t *supported_groups,
+        uint16_t *supported_groups_count_out, uint16_t max_count)
+{
+    POSIX_ENSURE_REF(supported_groups_count_out);
+    *supported_groups_count_out = 0;
+    POSIX_ENSURE_REF(ch);
+    POSIX_ENSURE_REF(supported_groups);
+
+    s2n_parsed_extension *supported_groups_extension = NULL;
+    POSIX_GUARD(s2n_client_hello_get_parsed_extension(S2N_EXTENSION_SUPPORTED_GROUPS, &ch->extensions, &supported_groups_extension));
+
+    struct s2n_stuffer extension_stuffer = { 0 };
+    POSIX_GUARD(s2n_stuffer_init_written(&extension_stuffer, &supported_groups_extension->extension));
+
+    uint16_t supported_groups_count = 0;
+    POSIX_GUARD_RESULT(s2n_client_supported_groups_parse_groups_count(&extension_stuffer, &supported_groups_count));
+    POSIX_ENSURE_LTE(supported_groups_count, max_count);
+
+    for (size_t i = 0; i < supported_groups_count; i++) {
+        POSIX_GUARD(s2n_stuffer_read_uint16(&extension_stuffer, &supported_groups[i]));
+    }
```

Comment:
Is there any particular reason we do this parsing? Why not just copy the memory over entirely / call s2n_client_hello_get_extension_by_id ourselves once we've validated the length? We're not actually doing anything with the uint16s we're reading. Just copying them.

Reply:
Don't we need `s2n_stuffer_read_uint16()` to write to the buffer with the correct endianness? My unit tests were failing when I tried to just copy it over.